### DISCUSSION TITLE
Cross compile for armv7-unknown-linux-gnueabihf

### DISCUSSION THREAD

**Discussion by eliabieri** [gyscos/cursive#684] [Q&A] [answered]
I try to cross compile for `armv7-unknown-linux-gnueabihf` but I get the following linker error:
```bash
w" "-Wl,-O1" "-nodefaultlibs"
  = note: /usr/lib/gcc-cross/arm-linux-gnueabihf/9/../../../../arm-linux-gnueabihf/bin/ld: cannot find -lncursesw
          /usr/lib/gcc-cross/arm-linux-gnueabihf/9/../../../../arm-linux-gnueabihf/bin/ld: cannot find -ltinfo
          collect2: error: ld returned 1 exit status
```
I've installed the `libncurses5-dev` package.

Steps to reproduce:
1) Install [Cross](https://github.com/cross-rs/cross)
2) Clone `cursive`
3) `cd cursive/cursive/examples/`
4) Create `Cross.toml` with the following content:
```toml
[build]
pre-build = ["apt-get install libncurses5-dev -y"]
```
5) Run `cross build --target armv7-unknown-linux-gnueabihf`

**Comment by gyscos** [comment, marked as answer]
Hi, and thanks for the question!

The ncurses backend is currently not the easiest one to cross-compile.
You'll need the development libraries installed for the target architecture (which is why `-lncursesw` and `-ltinfo` don't work in your case, you need these libraries available to the cross compiler), but even after that, as you noticed the `ncurses-sys` crate currently compiles and runs a tiny binary at build-time to check some type sizes, which is not easy when you compile for a different architecture than you run.

In the short-term I would recommend trying a different backend like `crossterm`, which should cross-compile more easily.

**Reply by eliabieri** [reply]
Thank you very much for your detailed answer!
I'm actually not using the ncurses backend, but it still seems to be required during compilation of `cursive`.  
This is the content of my `Cargo.toml`:
```toml
cursive = { version = "0.20", features = ["termion-backend"] }
```

**Reply by gyscos** [reply]
You would need to disable the default features which includes ncurses (`default-features = false`)

**Reply by eliabieri** [reply]
Oh, that so good to know!
Maybe would be worth to be added to the docs as others maybe also don't understand this.

**Reply by gyscos** [reply]
The first page includes a link to the wiki where this is mentioned:
https://github.com/gyscos/cursive/wiki/Backends

But I agree, it could be made more visible, possibly in the lib's top-level documentation.

**Reply by eliabieri** [reply]
Oh I see. Sorry didn't see that one.  
Now I can finally cross compile `cursive`!

**Comment by michalfita** [comment]
On Ubuntu 23.04, w/o `cargo-cross`:
```
sudo dpkg --add-architecture arm64
sudo apt install gcc-aarch64-linux-gnu binutils-aarch64-linux-gnu libncurses-dev:arm64
cat - <<EOF >> .cargo/config
[target.aarch64-unknown-linux-gnu]
linker = "aarch64-linux-gnu-gcc"
EOF
cargo --target aarch64-linux-gnu-gcc
```
and it builds nicely with `crossterm`.
